File: ApnsExample/ApnsExampleObjC/ViewController.m
//
//  ViewController.m
//  ApnsExampleObjC
//
//  Created by Pranav Prakash on 14/03/16.
//  Copyright © 2016 Aurora Borialis. All rights reserved.
//

#import "ViewController.h"
#import "OctoAPI.h"

@interface ViewController ()

@end

@implementation ViewController

UILabel* myLabel;

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    /*
     Create Buttons for various calls
     */
    UIButton *button = [UIButton buttonWithType:UIButtonTypeCustom];
    [button addTarget:self
               action:@selector(appLogoutBtnHandler:)
     forControlEvents:UIControlEventTouchUpInside];
    [button setTitle:@"App Logout" forState:UIControlStateNormal];
    UIColor *color = [UIColor colorWithRed:14.0/255.0 green:114.0/255.0 blue:199.0/255.0 alpha:1];
    [button setBackgroundColor:color];
    button.frame = CGRectMake(10, 100, 200, 50);
    [self.view addSubview:button];
    
    
    UIButton *button2 = [UIButton buttonWithType:UIButtonTypeCustom];
    [button2 addTarget:self
                action:@selector(appLoginBtnHandler:)
      forControlEvents:UIControlEventTouchUpInside];
    [button2 setTitle:@"App Login" forState:UIControlStateNormal];
    [button2 setBackgroundColor:[UIColor colorWithRed:119.0/255.0 green:114.0/255.0 blue:199.0/255.0 alpha:1]];
    button2.frame = CGRectMake(10, 200, 200, 50);
    [self.view addSubview:button2];
    
    UIButton *button3 = [UIButton buttonWithType:UIButtonTypeCustom];
    [button3 addTarget:self
                action:@selector(pageViewBtnHandler:)
      forControlEvents:UIControlEventTouchUpInside];
    [button3 setTitle:@"Page View" forState:UIControlStateNormal];
    [button3 setBackgroundColor:[UIColor colorWithRed:119.0/255.0 green:9.0/255.0 blue:199.0/255.0 alpha:1]];
    button3.frame = CGRectMake(10, 300, 200, 50);
    [self.view addSubview:button3];
    
    UIButton *button4 = [UIButton buttonWithType:UIButtonTypeCustom];
    [button4 addTarget:self
                action:@selector(productPageViewBtnHandler:)
      forControlEvents:UIControlEventTouchUpInside];
    [button4 setTitle:@"Product Page View" forState:UIControlStateNormal];
    [button4 setBackgroundColor:[UIColor colorWithRed:119.0/255.0 green:114.0/255.0 blue:1.0/255.0 alpha:1]];
    button4.frame = CGRectMake(10, 400, 200, 50);
    [self.view addSubview:button4];
    
    
    myLabel =  [[UILabel alloc] initWithFrame: CGRectMake(10, 455, 300, 100)];
    myLabel.text = @"Responses show here...";
    myLabel.numberOfLines = 3;
    [self.view addSubview:myLabel];

    
}

- (void)locationManager:(CLLocationManager *)manager
     didUpdateLocations:(NSArray *)locations {
    CLLocation *location = [locations lastObject];
    NSLog(@"lat%f - lon%f", location.coordinate.latitude, location.coordinate.longitude);
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}



- (void) appLogoutBtnHandler: (UIButton*)sender {
    OctoAPI *api = [[OctoAPI alloc] init];
    [api sendAppLogoutCall:2  onCompletion: ^(NSString* response){
        NSLog(@"Got response from App.Logout %@", response);
        dispatch_async(dispatch_get_main_queue(), ^{
            [myLabel setText:response];
        });
        
    }];
}

- (void) appLoginBtnHandler: (UIButton*)sender {
    OctoAPI *api = [[OctoAPI alloc] init];
    [api sendAppLoginCall:2  onCompletion: ^(NSString* response){
        NSLog(@"Got response from App.Login %@", response);
        dispatch_async(dispatch_get_main_queue(), ^{
            [myLabel setText:response];
        });
    }];
}

- (void) pageViewBtnHandler: (UIButton*)sender {
    OctoAPI *api = [[OctoAPI alloc] init];
    [api sendPageViewCall:2 routeUrl: @"Home#Index"
               categories:@[@"Aldo", @"Women"]
                     tags:@[@"Red", @"Handbag", @"Leather"]
             onCompletion: ^(NSString* response){
        NSLog(@"Got response from Page.View %@", response);
        dispatch_async(dispatch_get_main_queue(), ^{
            [myLabel setText:response];
        });
    }];
}

- (void) productPageViewBtnHandler: (UIButton*)sender {
    OctoAPI *api = [[OctoAPI alloc] init];
    [api sendProductPageViewCall:2
                        routeUrl:@"Home#Deals"
                       productId:88
                     productName:@"SmartPhone"
                           price:899.9
                      categories:@[@"Electronics", @"Mobile"]
                            tags:@[@"Delhi", @"Motorola"]
                    onCompletion: ^(NSString* response){
                            NSLog(@"Got response from Productpage.view %@", response);
                        dispatch_async(dispatch_get_main_queue(), ^{
            [myLabel setText:response];
        });
    }
     ];
}


@end

File: ApnsExample/ApnsExample/ViewController.swift
//
//  ViewController.swift
//  ApnsExample
//
//  Copyright © 2016 Aurora Borialis. All rights reserved.
//

import UIKit
import CoreLocation

class ViewController: UIViewController, CLLocationManagerDelegate {
    
    var locationManager: CLLocationManager!
    
    var label = UILabel(frame: CGRectMake(10, 460, 300, 50))


    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        initializeNotificationServices()
        
        locationManager = CLLocationManager()
        locationManager.delegate = self
        locationManager.requestWhenInUseAuthorization()
        
        // Add buttons for Events
        
        // add app login button
        let button   = UIButton(type: UIButtonType.System) as UIButton
        button.frame = CGRectMake(10, 100, 200, 50)
        button.backgroundColor = UIColor.greenColor()
        button.setTitle("App Login", forState: UIControlState.Normal)
        button.addTarget(self, action: "AppLoginBtnHandler:", forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(button)
        
        // add app logout button
        let button2   = UIButton(type: UIButtonType.System) as UIButton
        button2.frame = CGRectMake(10, 200, 200, 50)
        button2.backgroundColor = UIColor.blueColor()
        button2.setTitle("App Logout", forState: UIControlState.Normal)
        button2.addTarget(self, action: "AppLogoutBtnHandler:", forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(button2)
        
        // add page view button
        let button3   = UIButton(type: UIButtonType.System) as UIButton
        button3.frame = CGRectMake(10, 300, 200, 50)
        button3.backgroundColor = UIColor.redColor()
        button3.setTitle("Page View", forState: UIControlState.Normal)
        button3.addTarget(self, action: "PageViewBtnHandler:", forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(button3)
        
        // add product page view button
        let button4   = UIButton(type: UIButtonType.System) as UIButton
        button4.frame = CGRectMake(10, 400, 200, 50)
        button4.backgroundColor = UIColor.purpleColor()
        button4.setTitle("Product Page View", forState: UIControlState.Normal)
        button4.addTarget(self, action: "ProductPageViewBtnHandler:", forControlEvents: UIControlEvents.TouchUpInside)
        self.view.addSubview(button4)
        
        // Create a label for showing Response IDs
        label.text = "Events Results here..."
        label.numberOfLines = 3
        self.view.addSubview(label)
        
        
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    func initializeNotificationServices() -> Void {
        let settings = UIUserNotificationSettings(forTypes: [.Sound, .Alert, .Badge], categories: nil)
        UIApplication.sharedApplication().registerUserNotificationSettings(settings)
        
        // This is an asynchronous method to retrieve a Device Token
        // Callbacks are in AppDelegate.swift
        // Success = didRegisterForRemoteNotificationsWithDeviceToken
        // Fail = didFailToRegisterForRemoteNotificationsWithError
        UIApplication.sharedApplication().registerForRemoteNotifications()
    }
    
    func showAPIResult(response: NSString) -> (Void) {
        print(response)
        dispatch_async(dispatch_get_main_queue()) { () -> Void in
            self.label.text = response as String
        }
        
    }
    
    func AppLoginBtnHandler(sender:UIButton!){
        let api = OctoAPI()
        api.sendAppLoginCall(5) { (result) -> Void in
            self.showAPIResult(result)
        }
    }
    
    
    func AppLogoutBtnHandler(sender:UIButton!){
        let api = OctoAPI()
        api.sendAppLogoutCall(5) { (result) -> Void in
            self.showAPIResult(result)
        }
    }
    
    
    func PageViewBtnHandler(sender:UIButton) {
        
        let api = OctoAPI()
        
        let routeUrl = "/Home"
        let categories = ["something", "something else"]
        let tags = ["cat1", "cat2"]
        
        api.sendPageViewCall(5, routeUrl: routeUrl,
            categories: categories, tags: tags) { (result) -> Void in
                self.showAPIResult(result)
        }
    }
    
    
    func ProductPageViewBtnHandler(sender:UIButton) {
        
        let api = OctoAPI()
        
        let userId = 5
        let routeUrl = "/Home/Phone"
        let productName = "Smartphone Series S02"
        let price = 999.00
        let productId = 635373
        let categories = ["electronics", "phones"]
        let tags = ["selfie", "cheap"]
        
        api.sendProductPageViewCall(userId, routeUrl: routeUrl,
            productId: productId, price: price, productName: productName,
            categories: categories, tags: tags) { (result) -> Void in
                self.showAPIResult(result)
        }
    }
    
    func locationManager(manager: CLLocationManager, didUpdateLocations locations: [CLLocation]) {
        OctoAPI.updateLocation(locations)
    }
    
    func locationManager(manager: CLLocationManager, didFailWithError error: NSError) {
        print("didFailWithError: \(error.description)")

    }
    func locationManager(manager: CLLocationManager, didChangeAuthorizationStatus status: CLAuthorizationStatus) {
        if status == .AuthorizedWhenInUse {
            locationManager.startUpdatingLocation()
        }
        if status == .AuthorizedAlways {
            if CLLocationManager.isMonitoringAvailableForClass(CLBeaconRegion.self) {
                if CLLocationManager.isRangingAvailable() {
                    // do stuff
                }
            }
        }
    }
}